src/libraries/System.Net.Quic/src/System/Net/Quic/Implementations/MsQuic/Interop/msquic.cs

```diff
@@ -0,0 +1,191 @@
+#pragma warning disable IDE0073
+//
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+//
+#pragma warning restore IDE0073
+
+using System;
+using System.Diagnostics;
+using System.Net.Quic;
+using System.Runtime.CompilerServices;
+using System.Runtime.InteropServices;
+
+namespace Microsoft.Quic
+{
+    internal unsafe partial struct QUIC_BUFFER
+    {
+        public Span<byte> Span => new(Buffer, (int)Length);
+    }
+
+    internal partial class MsQuic
+    {
+        public static unsafe QUIC_API_TABLE* Open()
+        {
+            QUIC_API_TABLE* ApiTable;
+            int Status = MsQuicOpenVersion(2, (void**)&ApiTable);
+            ThrowIfFailure(Status);
+            return ApiTable;
+        }
+
+        public static unsafe void Close(QUIC_API_TABLE* ApiTable)
+        {
+            MsQuicClose(ApiTable);
+        }
+
+        public static void ThrowIfFailure(int status, string? message = null)
+        {
+            if (StatusFailed(status))
+            {
+                // TODO make custom exception, and maybe throw helpers
+                throw new MsQuicException(status, message);
+            }
+        }
+
+        public static bool StatusSucceeded(int status)
+        {
+            if (RuntimeInformation.IsOSPlatform(OSPlatform.Windows))
+            {
+                return status >= 0;
+            }
+            else
+            {
+                return status <= 0;
+            }
+        }
+
+        public static bool StatusFailed(int status)
+        {
+            if (RuntimeInformation.IsOSPlatform(OSPlatform.Windows))
+            {
+                return status < 0;
+            }
+            else
+            {
+                return status > 0;
+            }
+        }
+
+        public static readonly int QUIC_STATUS_SUCCESS = RuntimeInformation.IsOSPlatform(OSPlatform.Windows) ? MsQuic_Windows.QUIC_STATUS_SUCCESS : RuntimeInformation.IsOSPlatform(OSPlatform.OSX) ? MsQuic_MacOS.QUIC_STATUS_SUCCESS : MsQuic_Linux.QUIC_STATUS_SUCCESS;
```

Comment:
```suggestion
        public static int QUIC_STATUS_SUCCESS => OperatingSystem.IsWindows() ? MsQuic_Windows.QUIC_STATUS_SUCCESS : OperatingSystem.IsMacOS() ? MsQuic_MacOS.QUIC_STATUS_SUCCESS : MsQuic_Linux.QUIC_STATUS_SUCCESS;
```
With the new more efficient way to check for OS using `OperatingSystem.IsXXX`, it is more efficient to define these constants as properties. It will allow inlining the right constants directly into JITed code and avoid the large static constructor.

Reply:
Sounds good, I'll make that change on our end too.

Reply:
@jkotas is a property still good if one of the conditions is multiple checks or'd together? The current line for status checks now is 

public static readonly int QUIC_STATUS_SUCCESS = OperatingSystem.IsWindows() ? MsQuic_Windows.QUIC_STATUS_SUCCESS : (OperatingSystem.IsLinux() || OperatingSystem.IsAndroid()) ? MsQuic_Linux.QUIC_STATUS_SUCCESS : MsQuic_MacOS.QUIC_STATUS_SUCCESS;

And then I have a check for bsd os's with 

if (OperatingSystem.IsFreeBSD() || OperatingSystem.IsIOS() || OperatingSystem.IsMacOS() || OperatingSystem.IsMacCatalyst() || OperatingSystem.IsTvOS() || OperatingSystem.IsWatchOS())

later on.

Reply:
You may want to mark these properties with `[MethodImpl(MethodImplOptions.AggressiveInlining)` so that the JIT does not give up inlining the method due to the more complex condition.

I think the property is still better than readonly field. The JIT should have no problem with turning the property into a constant.

Reply:
Is allowing MethodImpl on a property a .NET 7 thing? I remember seeing some work being done on allowing attributes on properties, but don't remember which version, and .NET 6 is not allowing it on a property.

Reply:
Right, it needs to go on the getter, like this:
`int MyProperty { [MethodImpl(MethodImplOptions.AggressiveInlining)] get => 1; }`